### PR TITLE
26 implement substitution alternative

### LINKED ISSUE #26: Implement substitution alternative
A first implementation of the substitution algorithm using only merge. Implement as follows

- For all composite propositions `x` with exactly one sub propositions y, set `x = y`
- For all composite propositions with exactly two sub propositions x and y, try `x.merge(y)` and if fail create new variable.

### PR DESCRIPTION
**Pull request by znittzel**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -432,6 +443,12 @@ impl Clone for Variable {
     }
 }
 
+impl PartialEq for Variable {
```
Comment (ourmoa) [inline src/lib.rs:417]: More descriptive function name? 
```suggestion
impl IdEq for Variable {
```

File: src/lib.rs [hunk 2]
```diff
@@ -620,48 +714,563 @@ impl Theory {
...
+                let indices_ok = ae.0.indices == ae.1.indices;
+                let coeffs_ok = ae.0.coeffs == ae.1.coeffs;
+                return bias_ok && bounds_ok && indices_ok && coeffs_ok;
+            });
+        }
+
+        // Depth 3
+        // 0: 1 & 2
+        // 1: 3 + 4 >= 6 (3 has bounds (-5,5), 4 is bool) 
+        // 2: 5 + 6 >= 5 (5 has bounds (-3,3), 6 is bool)
```
Comment (ourmoa) [inline src/lib.rs]:
```suggestion
        // 2: 5 + 6 >= 4 (5 has bounds (-3,3), 6 is bool) 
```

**Review by ourmoa — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (znittzel) [reply src/lib.rs:417]: Isn't PartialEq a Trait?

File: src/lib.rs [hunk 3]
```diff
@@ -6,7 +6,9 @@ use std::hash::Hash;
 use std::{collections::HashMap};
 use std::fmt::Display;
 use itertools::iproduct;
```
Comment (ourmoa) [inline src/lib.rs]:
```suggestion
```

File: src/lib.rs [hunk 4]
```diff
@@ -6,7 +6,9 @@ use std::hash::Hash;
...
+use std::ops::Range;
```
Comment (ourmoa) [inline src/lib.rs]:
```suggestion
```

File: src/lib.rs [hunk 5]
```diff
@@ -6,7 +6,9 @@ use std::hash::Hash;
...
 use std::cmp;
+use itertools::Itertools;
```
Comment (ourmoa) [inline src/lib.rs]:
```suggestion
```

File: src/lib.rs [hunk 6]
```diff
@@ -470,200 +469,891 @@ pub struct AtLeast {
...
+    /// ```
+    fn to_lineq_extended(&self, given_id: u32, variable_hm: &HashMap<u32, Variable>) -> GeLineq {
+        return GeLineq::merge_disj(
+            &(variable_hm.get(&given_id).expect(&format!("variable id `{given_id}` does not exist in variable hash map"))).to_lineq_neg(),
+            &self.to_lineq(variable_hm)
+        ).expect("could not merge disjunctions") // <- should always be able to merge here
+    }
+
+    /// Transforms into a linear inequality constraint. If sub constraints fulfills certain requirements, then the sub constraints
+    /// will be merged into this constraint. If requirements is not met, then `to_lineq_extended` is called.
```
Comment (ourmoa) [inline src/lib.rs]:
```suggestion
    /// will be merged into this constraint. If the requirements are not met, then `to_lineq_extended` is called. 
```

File: src/lib.rs [hunk 7]
```diff
@@ -470,200 +469,891 @@ pub struct AtLeast {
...
-}
-
-/// A `Theory` is a list of statements with a common name (id).
-pub struct Theory {
-    pub id : String,
-    pub statements : Vec<Statement>
-}
+}
+
+/// A `Statement` is a declaration of a expression (or proposition) connected to a `Variable`.
```
Comment (ourmoa) [inline src/lib.rs]:
```suggestion
/// A `Statement` is a declaration of an expression (or proposition) connected to a `Variable`.
```

File: src/lib.rs [hunk 8]
```diff
@@ -470,200 +469,891 @@ pub struct AtLeast {
...
+                    id_curr = &statement.variable.id;
+                    break;
+                }
+            }
+            counter += 1;
+        }
+        return id_curr;
+    }
+
+    // All bottom / atomic variable ids. E.i. statements that don't have any children.
```
Comment (ourmoa) [inline src/lib.rs]:
```suggestion
    // All bottom / atomic variable ids, e.i. statements that don't have any children.
```

File: src/lib.rs [hunk 9]
```diff
@@ -470,200 +469,891 @@ pub struct AtLeast {
...
+        return lineqs;
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // Validates all combinations generated from `variable_bounds` in two vectors, lineqs0 och lineqs1, of GeLineq data types. 
+    // `all_relation` how the results from lineqs0 and lineqs1 relates. If `all_relation` is true then either both
```
Comment (ourmoa) [inline src/lib.rs]: feels like something is missing in the last sentence?

File: src/lib.rs [hunk 10]
```diff
@@ -470,200 +469,891 @@ pub struct AtLeast {
...
+        let actual: Vec<GeLineq> = t.to_lineqs(true, true);
+        let expected: Vec<GeLineq> = vec![
+            GeLineq {
+                bias: -34,
+                bounds: vec![(-5,5),(0,1),(-3,3),(0,1)],
+                coeffs: vec![5,5,1,1],
+                indices: vec![3,4,5,6]
+            },
+        ];
+        assert!(validate_theory_lineqs(t, actual, expected));
```
Comment (ourmoa) [inline src/lib.rs]: I believe we want to validate the theory based on the input and actual and not on the expected

**Review by ourmoa — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 11]
```diff
@@ -470,200 +450,820 @@ pub struct AtLeast {
...
+                    ).bounds
+                }
+            ).collect(),
             indices: self.ids.to_vec()
         }
     }
+
+    /// Transforms into a linear inequality constraint. Unlike `to_lineq` this is extended to include the given variable
+    /// such that it holds logic as "if `given_id` is true then `self` must be true".
+    /// ```
```
Comment (ourmoa) [inline src/lib.rs]:
```suggestion
```
Remove?

File: src/lib.rs [hunk 12]
```diff
@@ -470,200 +450,820 @@ pub struct AtLeast {
...
+                    if self.indices[j] == variables[i].0 {
+                        res = res + self.coeffs[j]*variables[i].1;
+                    }
+                }
+            }
+            return res + self.bias >= 0;
+        }
+    }
+
+    // Validates all combinations generated from `variable_bounds` in two vectors, lineqs0 och lineqs1, of GeLineq data types.
```
Comment (ourmoa) [inline src/lib.rs]:
```suggestion
    // Validates that all combinations, generated from `variable_bounds`, in two vectors of GeLineqs gives the same result, i.e. they hold the same truth table```

File: src/lib.rs [hunk 13]
```diff
@@ -470,200 +450,820 @@ pub struct AtLeast {
...
+                        res = res + self.coeffs[j]*variables[i].1;
+                    }
+                }
+            }
+            return res + self.bias >= 0;
+        }
+    }
+
+    // Validates all combinations generated from `variable_bounds` in two vectors, lineqs0 och lineqs1, of GeLineq data types. 
+    // `all_relation` how the results from lineqs0 and lineqs1 relates. If `all_relation` is true then either both
```
Comment (ourmoa) [inline src/lib.rs]:
```suggestion
    // The `all_relation` states how the GeLineq within each vector relates. If `all_relation` is true then all GeLineqs of the vector must hold to produce `True` as output. If `all_relation` is false then it is treated as an any, and only one GeLineq of the vector must hold to produce `True` as output. 
```

**Review by ourmoa — CHANGES_REQUESTED** [review]

**Review by ourmoa — APPROVED** [review]

**Pull request merged**